### PR TITLE
Integrate Elyra into Open Data Hub

### PR DESCRIPTION
**Pull request by akchinSTC**
Add Elyra as a notebook option in Open Data Hub

Open Data Hub uses JupyterHub to launch notebooks in Openshift and this  
includes Elyra as an option. Initial implementation includes a new docker image
to configured to use a default kubeflow installation colocated on the same 
cluster as well as support for CRI-o based container runtime.

Should be merged in conjunction with : https://github.com/elyra-ai/kfp-notebook/pull/46
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: etc/docker/odh/setup-elyra-metadata.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  # Otherwise, use jupyterhub username and replace any special characters with a dash
+  COS_BUCKET=$(python -c 'import sys;print(sys.argv[1].translate ({ord(c): "-" for c in "!@#$%^&*()[]{};:,./<>?\|`~=_+"}))' "$JUPYTERHUB_USER_NAME")
+  export COS_BUCKET
+fi
+
+if [[ ! -f "$DEFAULT_RUNTIME_FILE" ]]; then
+  elyra-metadata install runtimes --schema_name=kfp \
+                                  --name=my_kfp \
+                                  --display_name="Default Kubeflow Pipeline Runtime" \
+                                  --api_endpoint=http://ml-pipeline-ui.kubeflow/pipeline \
```
Comment (kevin-bates) [inline etc/docker/odh/setup-elyra-metadata.sh:19]: Could this KFP instance require DEX authentication?  If so, seems like `api_username` and `api_password` arguments would be necessary.

File: docs/source/recipes/deploying-elyra-with-opendatahub.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+your workstation.
+- Open a proxy to your cluster `oc proxy &`, this will run in the background.
+- Navigate to the `Landing Page` in your browser. NOTE: Replace the`Project` name in the URL with your own 
+http://localhost:8001/api/v1/namespaces/INSERT_PROJECT_NAME/services/http:jupyterhub:8080/proxy
+
+
+## Using Elyra with Open Data Hub
+- In the JupyterHub landing/spawner page, ensure you set the following two environmental variables before starting a notebook
+
+| NAME | VALUES | OPTIONAL | DESCRIPTION |
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply docs/source/recipes/deploying-elyra-with-opendatahub.md:127]: updated with changes

File: docs/source/recipes/deploying-elyra-with-opendatahub.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+|---|---|---|---|
+|COS_BUCKET| A-Z, a-z, - | X | This will be the bucket that your artifacts will be sent to post notebook execution, by default the jupyterhub username will be the bucket name|
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply docs/source/recipes/deploying-elyra-with-opendatahub.md:129]: updated character replacement and doc to reflect this

File: etc/docker/odh/setup-elyra-metadata.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+# Assumptions are existing kubeflow installation is in the kubeflow namespace
+DEFAULT_RUNTIME_FILE=$(jupyter --data-dir)/metadata/runtimes/my_kfp.json
+
+if  [[ ! -z "$JUPYTERHUB_USER_NAME" ]]; then
+  # Use COS_BUCKET env variable set in spawner, if not set, use default
+  export COS_BUCKET=${COS_BUCKET:-default}
+else
+  # Otherwise, use jupyterhub username and replace any special characters with a dash
+  COS_BUCKET=$(python -c 'import sys;print(sys.argv[1].translate ({ord(c): "-" for c in "!@#$%^&*()[]{};:,./<>?\|`~=_+"}))' "$JUPYTERHUB_USER_NAME")
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply etc/docker/odh/setup-elyra-metadata.sh:11]: thanks kevin. I wasnt sure about how all the flavors of unix shells implements the `tr` or other string manipulation utilities. Issues seem to arise when locales come into play as well so I went with python instead.

[the remark this answers is not in the document]
Comment (kevin-bates) [reply etc/docker/odh/setup-elyra-metadata.sh:11]: Great point. Perhaps we can put it in a function?

File: elyra/pipeline/processor_kfp.py [hunk 5]
```diff
@@ -181,11 +184,16 @@ def _cc_pipeline(self, pipeline, pipeline_name):
...
         for operation in pipeline.operations.values():
+
             operation_artifact_archive = self._get_dependency_archive_name(operation)
 
             self.log.debug("Creating pipeline component :\n {op} archive : {archive}".format(
                            op=operation, archive=operation_artifact_archive))
 
+            if crio_container_runtime:
+                # Volume size to create when using CRI-o, NOTE: IBM Cloud minimum is 20Gi
+                crio_emptydir_volume_size = '20Gi'
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply elyra/pipeline/processor_kfp.py:195]: If in the future we want to change the size of the volume on a per notebook basis, for example a notebook that needs a larger workspace for intermediate and large data sets, we might want to keep this here until we can find a place to define the size like in common properties or something and take that value to replace this one or use this default value. Ideally we'd be able to calculate this value on the fly but havent thought of a good way of going about that right now

File: elyra/pipeline/processor_kfp.py [hunk 6]
```diff
@@ -161,6 +161,9 @@ def _cc_pipeline(self, pipeline, pipeline_name):
         cos_directory = pipeline_name
         cos_bucket = runtime_configuration.metadata['cos_bucket']
 
+        crio_emptydir_volume_size = ''
+        crio_container_runtime = os.getenv("CRIO_RUNTIME", False)
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply elyra/pipeline/processor_kfp.py:165]: thanks! took the boolean casting you used from EG

File: etc/docker/odh/setup-elyra-metadata.sh (hunk 1 above)
Comment (akchinSTC) [reply etc/docker/odh/setup-elyra-metadata.sh:19]: odh's kubeflow manifests dont appear to have a kfdef for openshift and istio(w/ dex enabled) yet. Need to check with them to see if they can support moving forward

File: etc/docker/odh/setup-elyra-metadata.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# Assumptions are existing kubeflow installation is in the kubeflow namespace
+DEFAULT_RUNTIME_FILE=$(jupyter --data-dir)/metadata/runtimes/my_kfp.json
+
+if  [[ ! -z "$JUPYTERHUB_USER_NAME" ]]; then
+  # Use COS_BUCKET env variable set in spawner, if not set, use default
+  export COS_BUCKET=${COS_BUCKET:-default}
+else
```
[the remark this answers is not in the document]
Comment (akchinSTC) [reply etc/docker/odh/setup-elyra-metadata.sh:9]: Had to revert back to using just the COS_BUCKET. JUPYTERHUB_USER_NAME is set much later after the spawner kicks off. Looking at alternatives now

**Pull request merged**